### DISCUSSION TITLE
Question about StagesSubject

### DISCUSSION THREAD

**Discussion by NickWu** [Autodesk/maya-usd#2516] [Q&A]
Hi, I have a question about the logic of the following code: 

https://github.com/Autodesk/maya-usd/blob/00e63f3dcdc96711c5009477296f1eac6ed39e41/lib/mayaUsd/ufe/StagesSubject.cpp#L581-L583

The comment on the above code says : We should have no listeners and stage map is dirty. So I wonder why not just returning from the function when `g_StageMap` is not dirty and `fStageListeners` is not empty? What is the reason to keep proceeding the rest of the logic when both statements of `TF_VERIFY` failed? Thank you.

**Comment by ppt-adsk** [comment]
Hi @NickWu, given that this is essentially an assertion, the software is in an unspecified state at this point.  TF_AXIOM would cause a program exit, which is too strong.  My opinion is at this point the user should be in a state where they can save their work, and that the software should be in a reasonable enough state to allow this to succeed.  Your suggestion has merits; you could try it out to see if the plugin remains stable in such a case.  However, my hope is that you will never hit this code path :)

Is this causing you problems, or are you simply trying to get a better understanding of the code?  Cheers.

**Reply by NickWu** [reply]
Thank you for your reply! 

Yes, at Animal Logic, we found that `outStageData` attribute (derived from `MayaUsdProxyShapeBase`) is being requested to be recomputed many times when moving a vehicle rig (rigged with math nodes connections) in Maya which drives the prims in usd file loaded in AL `ProxyShape `which is derived from `MayaUsdProxyShapeBase`:
https://github.com/Autodesk/maya-usd/blob/00e63f3dcdc96711c5009477296f1eac6ed39e41/plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/ProxyShape.cpp#L1646-L1648

In turn, `computeOutStageData()` would be called which would send `MayaUsdProxyStageSetNotice`:
https://github.com/Autodesk/maya-usd/blob/00e63f3dcdc96711c5009477296f1eac6ed39e41/plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/ProxyShape.cpp#L1584

which would call `StagesSubject::onStageSet` many times over the course of moving a vehicle rig: 
https://github.com/Autodesk/maya-usd/blob/00e63f3dcdc96711c5009477296f1eac6ed39e41/lib/mayaUsd/ufe/StagesSubject.cpp#L562

As a result, we are seeing lots of error message being emitted in Maya script editor: 
```
// Error: Failed verification: ' g_StageMap.isDirty() ' // 
// Error: Failed verification: ' fStageListeners.empty() ' // 
// Error: Failed verification: ' g_StageMap.isDirty() ' // 
// Error: Failed verification: ' fStageListeners.empty() ' // 
// Error: Failed verification: ' g_StageMap.isDirty() ' // 
// Error: Failed verification: ' fStageListeners.empty() ' // 
// Error: Failed verification: ' g_StageMap.isDirty() ' // 
// Error: Failed verification: ' fStageListeners.empty() ' // 
```
For some reason, this issue is only happening starting from `Maya 2022.3.2-22.0.PFIX`. We don't see this issue in `Maya 2022.3.2-19.0.PFIX`

Even though I haven't figured out what is triggering this issue and why it's only happening in `Maya 2022.3.2-22.0.PFIX` and newer Maya 2022 versions, but if we could return from the `StagesSubject::onStageSet` function when `g_StageMap` is not dirty and `fStageListeners` is not empty if it's not necessary to proceed the result of the code, we could avoid this constant error message being emitted. So I am suggesting instead of using TF_VERIFY, we could do: 

```C++
if(!g_StageMap.isDirty() && !fStageListeners.empty())
    return;
```
One thing I am not sure about this approach is that if we do early return from the function like the question I mentioned, should we set `stageSetGuardCount=false` because the following code would set `stageSetGuardCount=true` when the if statement evaluates true: 
https://github.com/Autodesk/maya-usd/blob/00e63f3dcdc96711c5009477296f1eac6ed39e41/lib/mayaUsd/ufe/StagesSubject.cpp#L580

And since we are early returning from the function, we don't set `stageSetGuardCount` back to false like the end of the function does: 
https://github.com/Autodesk/maya-usd/blob/00e63f3dcdc96711c5009477296f1eac6ed39e41/lib/mayaUsd/ufe/StagesSubject.cpp#L606

What do you think? Thanks!
